source/common/http/filter_manager.cc

```diff
@@ -1284,6 +1298,14 @@ void FilterManager::encodeHeaders(ActiveStreamEncoderFilter* filter, ResponseHea
 
   const bool modified_end_stream = (end_stream && continue_data_entry == encoder_filters_.end());
   state_.non_100_response_headers_encoded_ = true;
+  if (allow_upstream_half_close_) {
+    const uint64_t response_status = Http::Utility::getResponseStatus(headers);
+    if (!(Http::CodeUtility::is2xx(response_status) || Http::CodeUtility::is1xx(response_status))) {
+      // Even if upstream half close is enabled the stream is closed on error responses from the
+      // server.
+      state_.force_close_stream_ = true;
```

Comment:
this seems like a really weird place to put this.  why not on all local resets and or local replies?

Reply:
I could be overthinking here. My thought was to allow independent half close only when server response is 2xx. If it is error (ore redirect) response Envoy fully closes the stream. This bit of code treats upstream error response as a `sendLocalReply`.

I was thinking about case where upstream response with error to a POST with large body. Presently Envoy 'helps' gracefully stop body upload. With independent half closes the upstream will have to follow up an error response with a reset, otherwise upload will continue. 

I'm not sure if Envoy needs to do this. We could get rid of this case and make it so independent half close is supported regardless of the upstream response status. And the perhaps add a flag if Envoy needs to support this corner case. Let me know what you think.